Add support for temporal types to batch importer

henriknyman opened the pull request:
- Date
- Time
- DateTime
- LocalTime
- LocalDateTime
- Duration

The default time zone for `Time` and `DateTime` is derived from the
database setting `dbms.db.timezone`.

henriknyman committed:
Add support for temporal types to batch importer

- Date
- Time
- DateTime
- LocalTime
- LocalDateTime
- Duration

The default time zone for `Time` and `DateTime` is derived from the
database setting `dbms.db.timezone`.

sherfert commented on community/csv/src/main/java/org/neo4j/csv/reader/Extractors.java:
```diff
@@ -952,6 +1008,186 @@ public AnyValue value()
         }
     }
 
+    public static class DateExtractor extends AbstractSingleValueExtractor<AnyValue>
```
It looks to me as if PointExtractor and all new Temporal extractors could share the field `value` and the methods `clear()` and `value()`. Maybe create an abstract type `AbstractSingleAnyValueExtractor extends AbstractSingleValueExtractor<AnyValue>` ?

sherfert commented on community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputBatchImportIT.java:
```diff
@@ -177,7 +186,14 @@ public int bufferSize()
+            node.property( "point", "\"   { x : -4.2, y : " + i + ", crs: WGS-84 } \"" );
+            node.property( "date", LocalDate.of( 2018, i % 12 + 1, i % 28 + 1 ) );
+            node.property( "time", OffsetTime.of( 1, i % 60, 0, 0, ZoneOffset.ofHours( 9 ) ) );
+            node.property( "dateTime",
```
I think we should have a dateTime without explicit timezone, too?

sherfert commented on community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputBatchImportIT.java:
```diff
@@ -482,10 +504,26 @@ private void buildUpExpectedData(
-                    assertEquals( expectedValue, actualValue );
-                };
+                    // Since there is no straightforward comparison for TemporalAmount we add it to a reference
+                    // point in time and compare the result
```
I don't understand this. DurationValues implement equals in a sensible way.

henriknyman replied:
But `DurationValue` is not a publicly exposed type. I am trying to use only what you would get out of the Core API for verification.

sherfert commented on community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputTest.java:
```diff
@@ -482,6 +488,156 @@ public void shouldParsePointPropertyValues() throws Exception
+        // GIVEN
+        DataFactory data = data(
+                ":ID,name,time:DateTime\n" +
+                "0,Mattias,2018-02-27T13:37\n" +
```
I think we need to have a case that _does_ specify an explicit timezone here, too.

sherfert commented on community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputTest.java:
```diff
@@ -482,6 +488,156 @@ public void shouldParsePointPropertyValues() throws Exception
+        DataFactory data = data(
+                ":ID,name,time:Time\n" +
+                "0,Mattias,13:37\n" +
+                "1,Johan,\"16:20:01\"\n" );
```
I think we need to have a case that does specify an explicit offset here, too.

henriknyman committed:
Fix some review comments

- Extract some shared code into AbstractSingleAnyValueExtractor
- Test that the default time zone is used
- Make the Temporal property verifier in the integration test
less implementation specific
- Test more variations of explicit/implicit time zone

henriknyman committed:
Fix codestyle...
